### DISCUSSION TITLE
Syntax question when attempting to pass params from a df to quarto_render

### DISCUSSION THREAD

**Discussion by ghost** [quarto-dev/quarto-cli#4327] [Q&A]
I am trying to take vars from a df and use them as params in a call to quarto_render:
```
library(tidyverse)
library(quarto)
one_person <- tibble(
  title=c("Rev"), 
  name=c("Albert")
)

test_tibble_2 <- tibble(
  input='index.qmd',
  output_format = "html",
  output_file = paste0("tibble2_", 
                       one_person$title, 
                       "_", 
                       one_person$name,
                       ".html"),
   execute_params = list(params=one_person$title)
)


pmap(
  test_tibble_2, 
  quarto_render
)



```
This approach works if I don't include execute_params (which was the point) but fails when I do.  My guess is that either  the 
one_person is not structured properly, or the execute_params is incorrectly specified. 

This is the failure I get; 

_Error in knit_params_get(input_lines, params) :    render params argument must be a named list Calls: .main -> execute -> <Anonymous> -> knit_params_get Execution halted Error in `pmap()`: ℹ In index: 1. Caused by error: ! System command 'quarto' failed Run `rlang::last_error()` to see where the error occurred.
--_

**Comment by mcanouil** [comment]
https://github.com/quarto-dev/quarto-cli/discussions/4323#discussioncomment-4950643

**Comment by ghost** [comment]
I am sorry, but this is not working for. 

Below I offer some code, in the form of 6 tests. Each tries to represent the smallest possible incremental change. 
````
library(tidyverse)
library(quarto)

# make a database of recipients

people <- tibble(
  title=c("Archbishop", "Brother"), 
  name=c("Albert", "Bob")
)

# an a 1 record version of people
one_person <- people %>% slice(1)

# test 1 Do one file,  with no use of map

quarto_render(
  input = "index.qmd",
  output_format = "html",
  output_file = "test_1_static_title.html",
  execute_params = list(title = "Sister", 
                        name =  "Sledge")
) 
````

Test1 works "Dear Sister Sledge" in the output.

Now, let's do the same  but from a tibble
Note use of one_person NOT people (so N=1, not N > 1)
````
test_2 <- tibble(
  input='index.qmd',
  output_format = "html",
  output_file = paste0("test_2_", 
                       one_person$title, 
                       "_", 
                       one_person$name,
                       ".html"),
#   execute_params = list(title=one_person$title, name=one_person$name)
)


pmap(
  test_2, 
  quarto_render
)
````

Test 2 works, but we didn't include the execute_params, but ... got the  output filename correct

````
# test 3
# note still a single case, using one_person df NOT N>1 df
# try again with just one execute_params

test_3<- tibble(
  input='index.qmd',
  output_format = "html",
  output_file = paste0("test_3_", 
                       one_person$title, 
                       "_", 
                       one_person$name,
                       ".html"),
  execute_params = list(title=one_person$title)
)


pmap(
  test_3, 
  quarto_render
)
````


Dies here "  render params argument must be a named list

R replies with: 

Fails: "params argument must be a named list

But  str(test_3) sure look like it is a named list


_tibble [1 × 4] (S3: tbl_df/tbl/data.frame)  
$ input         : chr "index.qmd"  
$ output_format : chr "html" 
$ output_file   : chr "test_3_Archbishop_Albert.html"  
$ execute_params:List of 1   ..$ title: chr "Archbishop"_
 

Test 4  Tries to add two params (but for a 1 record df)

````

test_4<- tibble(
  input='index.qmd',
  output_format = "html",
  output_file = paste0("test_4_", 
                       one_person$title, 
                       "_", 
                       one_person$name,
                       ".html"),
  execute_params = list(title=one_person$title, name=one_person$name)
)


pmap(
  test_4, 
  quarto_render
)
```

Test 5 Attempts to do this again for a df with N>1 rows

````

Test 5 tries to extend this to  one parameter, but multiple rows of input:

````
test_5<- tibble(
  input='index.qmd',
  output_format = "html",
  output_file = paste0("test_5_", 
                       people$title, 
                       "_", 
                       people$name,
                       ".html"),
  execute_params = list(title=people$title)
)

pmap(
  test_5, 
  quarto_render
)
````

Test   5 fails


Test 6

Here I am trying to use multiple rows, nd multiple parameters

````

test_6<- tibble(
  input='index.qmd',
  output_format = "html",
  output_file = paste0("test_6_", 
                       people$title, 
                       "_", 
                       people$name,
                       ".html"),
  execute_params = list(title=people$title, name=people$name)
)


pmap(
  test_6, 
  quarto_render
)
````

Test 6 fails, but that's not surprising. Need to fix test_3 first.

**Comment by mcanouil** [comment]
Here you don't provide the actual Quarto document you are using.
Without there is no way we can reproduce your errors and tell exactly what's wrong.

That said, the following reproducible examples work perfectly fine, thus there is a syntax error or another type of error in your code, which is unrelated to Quarto (R package or CLI).

```qmd
---
title: '`r params[["title"]]`'
params:
  title: "Default"
format: html
---

Text
```

```r
library(quarto)

lapply(
  X = head(letters),
  FUN = function(i) {
    quarto_render(
      input = "test.qmd",
      output_file = paste0("test_", i, ".html"),
      execute_params = list(
        title = paste("Test", toupper(i))
      )
    )
  }
)
```

```r
library(quarto)
library(purrr)
library(tibble)
pmap(
  .l = tibble(
    output_file = c("test_a.html", "test_b.html"),
    execute_params = list(
      list(title = "Test A"),
      list(title = "Test B")
    )
  ),
  .f = function(output_file, execute_params) {
    quarto_render(
      input = "test.qmd",
      output_file = output_file,
      execute_params = execute_params
    )
  }
)
```

```r
library(quarto)
library(purrr)
library(tibble)
pmap(
  .l = tibble(
    output_file = c("test_a.html", "test_b.html"),
    execute_params = list(
      list(title = "Test A"),
      list(title = "Test B")
    )
  ),
  .f = quarto_render,
  input = "test.qmd"
)
```

**Reply by ghost** [reply]
Thank you so much for the ideas and the generous effort.   You have been most kind. 

To the Quarto folks who I hope will read this,  this is not an obvious solution but is a pretty standard "mail merge" sort of problem. 

Syntax guidance to get multiple df  columns  (e.g. (name and title, for example)) "into" params values would be greatly appreciated. This perhaps could come in the form of a 'use case' or clearer documentation. 

Few of us have Mickaël Canouil's skills!

**Reply by mcanouil** [reply]
Which part of the documentation was not clear enough?
If you can target the exact part, it could be improved by maintainers. 😉

**Comment by ghost** [comment]
Mickael,  Your note was perfectly clear. What is needed, I am suggesting, is a more accessible way to get columns out of a df and into params, 

Your approach is technically lovely, but it would be easier for lay folk if the syntax was more accessible.
e.g
 
````
library(quarto)
library(purrr)
library(tibble)
#this is pseudo code; it does not run

pmap(
  .l = tibble(
    output_file = c("test_a.html", "test_b.html"),
    var1 = c("Dr", "Prof")),
     var2 = c("Dave", "Pat"), 
    execute_params = list(
# note that these are two variables, not two different values for same variable)
      list(title = some_helper_function(var1),
      list(name= some_helper_function(var_2)
    )
  ),
  .f = quarto_render,
  input = "test.qmd"
)
````
Ideally (for me) the pmap call would be _pmap(df, quarto_render)_ and all the other details would be addressed in structuring the input df. 

I really appreciate your assistance. I suppose another approach would be to writing run-specific yaml files to set the params, but that seemed unnecessarily indirect. 

Parameterized reports in Quarto will be a big deal, if we can do them easily from within the quarto_render function.

Again, my thanks and appreciation.

**Reply by mcanouil** [reply]
I don't follow ...

I wrote it as one block of code in my three examples, you only have to assign the tibble in `.x` into `df` from the third example.

There is no issue with the pmap or quarto_render function syntax.
The issue you are having would be the same for many other functions such as `rmarkdown::render`.
